File: bindings/ocaml/lib/safe_angle.ml
(* SPDX-License-Identifier: PMPL-1.0 *)
(* SPDX-FileCopyrightText: 2025 Hyperpolymath *)

(** Safe angle operations with unit conversions and normalization. *)

(** Angle unit types. *)
type unit_type =
  | Degrees
  | Radians
  | Gradians
  | Turns

(** Angle representation. *)
type t = {
  value: float;
  unit_type: unit_type;
}

(** Mathematical constants. *)
let pi = Float.pi
let tau = 2.0 *. pi

(** Create an angle in degrees. *)
let degrees value = { value; unit_type = Degrees }

(** Create an angle in radians. *)
let radians value = { value; unit_type = Radians }

(** Create an angle in gradians. *)
let gradians value = { value; unit_type = Gradians }

(** Create an angle in turns. *)
let turns value = { value; unit_type = Turns }

(** Convert angle to radians. *)
let to_radians angle =
  match angle.unit_type with
  | Radians -> angle.value
  | Degrees -> angle.value *. pi /. 180.0
  | Gradians -> angle.value *. pi /. 200.0
  | Turns -> angle.value *. tau

(** Convert angle to degrees. *)
let to_degrees angle =
  match angle.unit_type with
  | Degrees -> angle.value
  | Radians -> angle.value *. 180.0 /. pi
  | Gradians -> angle.value *. 0.9
  | Turns -> angle.value *. 360.0

(** Convert angle to gradians. *)
let to_gradians angle =
  match angle.unit_type with
  | Gradians -> angle.value
  | Degrees -> angle.value /. 0.9
  | Radians -> angle.value *. 200.0 /. pi
  | Turns -> angle.value *. 400.0

(** Convert angle to turns. *)
let to_turns angle =
  match angle.unit_type with
  | Turns -> angle.value
  | Degrees -> angle.value /. 360.0
  | Radians -> angle.value /. tau
  | Gradians -> angle.value /. 400.0

(** Convert angle to a specific unit. *)
let convert target_unit angle =
  let rad = to_radians angle in
  match target_unit with
  | Radians -> { value = rad; unit_type = Radians }
  | Degrees -> { value = rad *. 180.0 /. pi; unit_type = Degrees }
  | Gradians -> { value = rad *. 200.0 /. pi; unit_type = Gradians }
  | Turns -> { value = rad /. tau; unit_type = Turns }

(** Normalize angle to [0, 360) degrees or equivalent. *)
let normalize angle =
  let turns = to_turns angle in
  let normalized = turns -. Float.floor turns in
  convert angle.unit_type { value = normalized; unit_type = Turns }

(** Normalize angle to [-180, 180) degrees or equivalent. *)
let normalize_signed angle =
  let turns = to_turns angle in
  let normalized = turns -. Float.floor (turns +. 0.5) in
  convert angle.unit_type { value = normalized; unit_type = Turns }

(** Add two angles. *)
let add a b =
  let result = to_radians a +. to_radians b in
  convert a.unit_type { value = result; unit_type = Radians }

(** Subtract two angles. *)
let sub a b =
  let result = to_radians a -. to_radians b in
  convert a.unit_type { value = result; unit_type = Radians }

(** Multiply angle by scalar. *)
let mul scalar angle =
  { angle with value = angle.value *. scalar }

(** Divide angle by scalar. *)
let div scalar angle =
  if scalar = 0.0 then None
  else Some { angle with value = angle.value /. scalar }

(** Negate angle. *)
let negate angle =
  { angle with value = -. angle.value }

(** Compare two angles (after normalization). *)
let compare a b =
  let a_turns = to_turns (normalize a) in
  let b_turns = to_turns (normalize b) in
  Float.compare a_turns b_turns

(** Check if two angles are equal (within epsilon). *)
let equal ?(epsilon=1e-10) a b =
  let a_rad = to_radians (normalize a) in
  let b_rad = to_radians (normalize b) in
  Float.abs (a_rad -. b_rad) < epsilon

(** Compute sine of angle. *)
let sin angle =
  Float.sin (to_radians angle)

(** Compute cosine of angle. *)
let cos angle =
  Float.cos (to_radians angle)

(** Compute tangent of angle. *)
let tan angle =
  Float.tan (to_radians angle)

(** Create angle from sine value. *)
let asin value =
  if value < -1.0 || value > 1.0 then None
  else Some { value = Float.asin value; unit_type = Radians }

(** Create angle from cosine value. *)
let acos value =
  if value < -1.0 || value > 1.0 then None
  else Some { value = Float.acos value; unit_type = Radians }

(** Create angle from tangent value. *)
let atan value =
  Some { value = Float.atan value; unit_type = Radians }

(** Create angle from y/x coordinates (atan2). *)
let atan2 ~y ~x =
  { value = Float.atan2 y x; unit_type = Radians }

(** Get the smallest angle between two angles. *)
let difference a b =
  let diff = normalize_signed (sub a b) in
  if to_degrees diff < 0.0 then negate diff else diff

(** Linear interpolation between two angles (takes shortest path). *)
let lerp t a b =
  let diff = normalize_signed (sub b a) in
  add a (mul t diff)

(** Check if angle is in range [start, end) going clockwise. *)
let is_between ~start ~stop angle =
  let start_turns = to_turns (normalize start) in
  let stop_turns = to_turns (normalize stop) in
  let angle_turns = to_turns (normalize angle) in
  if start_turns <= stop_turns then
    angle_turns >= start_turns && angle_turns < stop_turns
  else
    angle_turns >= start_turns || angle_turns < stop_turns

(** Format angle as string. *)
let format angle =
  let suffix = match angle.unit_type with
    | Degrees -> "deg"
    | Radians -> "rad"
    | Gradians -> "grad"
    | Turns -> "turn"
  in
  Printf.sprintf "%.6f%s" angle.value suffix

(** Format as degrees with degree symbol. *)
let format_degrees angle =
  Printf.sprintf "%.2f\xC2\xB0" (to_degrees angle)

(** Common angles. *)
let zero = degrees 0.0
let right_angle = degrees 90.0
let straight_angle = degrees 180.0
let full_rotation = degrees 360.0
let quarter_turn = turns 0.25
let half_turn = turns 0.5

(** Cardinal directions. *)
let north = degrees 0.0
let east = degrees 90.0
let south = degrees 180.0
let west = degrees 270.0

(** Convert compass bearing to mathematical angle. *)
let bearing_to_math angle =
  normalize (sub (degrees 90.0) angle)

(** Convert mathematical angle to compass bearing. *)
let math_to_bearing angle =
  normalize (sub (degrees 90.0) angle)

File: bindings/julia/src/SafeAngle.jl
# SPDX-License-Identifier: PMPL-1.0
# SPDX-FileCopyrightText: 2025 Hyperpolymath

"""
    SafeAngle

Safe angle operations with conversions between degrees, radians, and other units.
"""
module SafeAngle

export Angle, AngleUnit
export from_degrees, from_radians, from_gradians, from_turns
export to_degrees, to_radians, to_gradians, to_turns
export normalize, normalize_signed, complement, supplement
export angle_add, angle_sub, angle_mul, angle_div
export sin_safe, cos_safe, tan_safe
export is_acute, is_right, is_obtuse, is_straight, is_reflex
export angle_between

"""
    AngleUnit

Enumeration of angle units.
"""
@enum AngleUnit begin
    Degrees
    Radians
    Gradians
    Turns
end

const TWO_PI = 2.0 * π
const DEG_TO_RAD = π / 180.0
const RAD_TO_DEG = 180.0 / π
const GRAD_TO_RAD = π / 200.0
const RAD_TO_GRAD = 200.0 / π

"""
    Angle

An angle stored internally in radians.
"""
struct Angle
    radians::Float64
end

"""
    from_degrees(deg::Real) -> Angle

Create angle from degrees.
"""
from_degrees(deg::Real)::Angle = Angle(Float64(deg) * DEG_TO_RAD)

"""
    from_radians(rad::Real) -> Angle

Create angle from radians.
"""
from_radians(rad::Real)::Angle = Angle(Float64(rad))

"""
    from_gradians(grad::Real) -> Angle

Create angle from gradians (1 turn = 400 gradians).
"""
from_gradians(grad::Real)::Angle = Angle(Float64(grad) * GRAD_TO_RAD)

"""
    from_turns(turns::Real) -> Angle

Create angle from turns (1 turn = 360 degrees).
"""
from_turns(turns::Real)::Angle = Angle(Float64(turns) * TWO_PI)

"""
    to_degrees(a::Angle) -> Float64

Convert to degrees.
"""
to_degrees(a::Angle)::Float64 = a.radians * RAD_TO_DEG

"""
    to_radians(a::Angle) -> Float64

Convert to radians.
"""
to_radians(a::Angle)::Float64 = a.radians

"""
    to_gradians(a::Angle) -> Float64

Convert to gradians.
"""
to_gradians(a::Angle)::Float64 = a.radians * RAD_TO_GRAD

"""
    to_turns(a::Angle) -> Float64

Convert to turns.
"""
to_turns(a::Angle)::Float64 = a.radians / TWO_PI

"""
    normalize(a::Angle) -> Angle

Normalize to [0, 2π) range.
"""
function normalize(a::Angle)::Angle
    r = mod(a.radians, TWO_PI)
    r < 0 && (r += TWO_PI)
    Angle(r)
end

"""
    normalize_signed(a::Angle) -> Angle

Normalize to [-π, π) range.
"""
function normalize_signed(a::Angle)::Angle
    r = mod(a.radians + π, TWO_PI) - π
    Angle(r)
end

"""
    complement(a::Angle) -> Union{Angle, Nothing}

Get the complement (90° - angle). Returns nothing if angle > 90°.
"""
function complement(a::Angle)::Union{Angle, Nothing}
    normalized = normalize(a)
    normalized.radians > π/2 && return nothing
    Angle(π/2 - normalized.radians)
end

"""
    supplement(a::Angle) -> Union{Angle, Nothing}

Get the supplement (180° - angle). Returns nothing if angle > 180°.
"""
function supplement(a::Angle)::Union{Angle, Nothing}
    normalized = normalize(a)
    normalized.radians > π && return nothing
    Angle(π - normalized.radians)
end

"""
    angle_add(a::Angle, b::Angle) -> Angle

Add two angles.
"""
angle_add(a::Angle, b::Angle)::Angle = Angle(a.radians + b.radians)

"""
    angle_sub(a::Angle, b::Angle) -> Angle

Subtract two angles.
"""
angle_sub(a::Angle, b::Angle)::Angle = Angle(a.radians - b.radians)

"""
    angle_mul(a::Angle, scalar::Real) -> Angle

Multiply angle by scalar.
"""
angle_mul(a::Angle, scalar::Real)::Angle = Angle(a.radians * scalar)

"""
    angle_div(a::Angle, scalar::Real) -> Union{Angle, Nothing}

Divide angle by scalar. Returns nothing on division by zero.
"""
function angle_div(a::Angle, scalar::Real)::Union{Angle, Nothing}
    scalar == 0 && return nothing
    Angle(a.radians / scalar)
end

"""
    sin_safe(a::Angle) -> Float64

Compute sine of angle.
"""
sin_safe(a::Angle)::Float64 = sin(a.radians)

"""
    cos_safe(a::Angle) -> Float64

Compute cosine of angle.
"""
cos_safe(a::Angle)::Float64 = cos(a.radians)

"""
    tan_safe(a::Angle) -> Union{Float64, Nothing}

Compute tangent of angle. Returns nothing near π/2 + nπ.
"""
function tan_safe(a::Angle)::Union{Float64, Nothing}
    normalized = normalize(a)
    # Check if near 90° or 270°
    near_90 = abs(normalized.radians - π/2) < 1e-10
    near_270 = abs(normalized.radians - 3π/2) < 1e-10
    (near_90 || near_270) && return nothing

    result = tan(a.radians)
    isfinite(result) ? result : nothing
end

"""
    is_acute(a::Angle) -> Bool

Check if angle is acute (0° < a < 90°).
"""
function is_acute(a::Angle)::Bool
    r = normalize(a).radians
    r > 0 && r < π/2
end

"""
    is_right(a::Angle; tolerance::Float64=1e-10) -> Bool

Check if angle is a right angle (90°).
"""
function is_right(a::Angle; tolerance::Float64=1e-10)::Bool
    r = normalize(a).radians
    abs(r - π/2) < tolerance
end

"""
    is_obtuse(a::Angle) -> Bool

Check if angle is obtuse (90° < a < 180°).
"""
function is_obtuse(a::Angle)::Bool
    r = normalize(a).radians
    r > π/2 && r < π
end

"""
    is_straight(a::Angle; tolerance::Float64=1e-10) -> Bool

Check if angle is straight (180°).
"""
function is_straight(a::Angle; tolerance::Float64=1e-10)::Bool
    r = normalize(a).radians
    abs(r - π) < tolerance
end

"""
    is_reflex(a::Angle) -> Bool

Check if angle is reflex (180° < a < 360°).
"""
function is_reflex(a::Angle)::Bool
    r = normalize(a).radians
    r > π && r < TWO_PI
end

"""
    angle_between(from::Angle, to::Angle) -> Angle

Get the smallest angle between two angles (always positive, <= 180°).
"""
function angle_between(from::Angle, to::Angle)::Angle
    diff = normalize(Angle(to.radians - from.radians))
    diff.radians > π ? Angle(TWO_PI - diff.radians) : diff
end

# Operator overloads
Base.:+(a::Angle, b::Angle) = angle_add(a, b)
Base.:-(a::Angle, b::Angle) = angle_sub(a, b)
Base.:-(a::Angle) = Angle(-a.radians)
Base.:*(a::Angle, s::Real) = angle_mul(a, s)
Base.:*(s::Real, a::Angle) = angle_mul(a, s)
Base.:/(a::Angle, s::Real) = something(angle_div(a, s), error("Division by zero"))

# Comparison
Base.:(==)(a::Angle, b::Angle) = normalize(a).radians ≈ normalize(b).radians
Base.:<(a::Angle, b::Angle) = a.radians < b.radians

# Display
function Base.show(io::IO, a::Angle)
    deg = to_degrees(a)
    print(io, "Angle(", round(deg, digits=2), "°)")
end

end # module
